sematic/db/migrate.py

```diff
@@ -47,53 +55,242 @@ def migrate():
+
+    try:
+        up() if direction == MigrationDirection.UP else down()
+
+        with db().get_engine().begin() as conn:
+            _update_version(conn, version, direction)
+
+    except Exception as e:
+        down() if direction == MigrationDirection.UP else up()
+        raise e
+
+
+def _update_version(conn: Connection, version: str, direction: MigrationDirection):
+    statement = (
+        "INSERT INTO schema_migrations(version) values (:version)"
+        if direction == MigrationDirection.UP
+        else "DELETE FROM schema_migrations WHERE version = :version"
+    )
+
+    conn.execute(text(statement), version=version)
+
+
+def common_options(f):
+    options = [
+        click.option(
+            "--env",
+            "env",
+            type=click.STRING,
+            help="Environment in which to run migrations.",
+            default="local",
+        ),
+        click.option(
+            "--verbose", "verbose", is_flag=True, default=False, help="INFO log level"
+        ),
+    ]
+    return functools.reduce(lambda x, opt: opt(x), options, f)
+
+
+@click.group("migrate")
+@common_options
+def main(env: str, verbose: bool):
+    _apply_common_options(env, verbose)
+
+
+def _apply_common_options(env, verbose):
+    switch_env(env)
+
+    if verbose:
+        logging.basicConfig(level=logging.INFO)
+
+
+@main.command("up", short_help="Apply outstanding migrations")
+@common_options
+@click.option("--schema-file", "file", type=click.STRING, default="schema.sql")
+def _migrate_up(env: str, verbose: bool, file: str):
+    """
+    Migrate the DB to the latest version.
+    """
+    _apply_common_options(env, verbose)
+    # Separate function to be able to invoke it outside of click
+    migrate_up()
+    dump_schema(file)
+
+
+def migrate_up():
+    """
+    Migrate the DB to the latest version.
+    """
+    logging.info("Running migrations on {}".format(get_config().db_url))
+
+    versions = _get_current_versions()
 
     migration_files = _get_migration_files()
 
     for migration_file in migration_files:
         version = migration_file.split("_")[0]
+
         if version in versions:
             logging.info("Already applied {}".format(migration_file))
             continue
 
-        logging.info("Applying {}".format(migration_file))
+        _run_migration(migration_file, version, MigrationDirection.UP)
+
 
-        with open(
-            os.path.join(get_config().migrations_dir, migration_file), "r"
-        ) as file:
-            sql = file.read()
+@main.command("down", short_help="Revert last migration")
+@common_options
+@click.option("--schema-file", "file", type=click.STRING, default="schema.sql")
+def _migrate_down(env: str, verbose: bool, file: str):
+    """
+    Revert the last migration.
+    """
+    _apply_common_options(env, verbose)
+    # Separate function to be able to invoke it outside of click
+    migrate_down()
+    dump_schema(file)
```

Comment:
I wonder whether this should be in a `finally`? Is it better to snapshot the schema as-is in the DB if a migrate fails, or as-is in the version control committed schema. Probably the latter, but let me know what you think.

Reply:
Yeah we should probably keep the schema file as source of truth for the canonical schema.